### PR TITLE
Add block_number to Contract methods

### LINKED ISSUE #823: Support `block_number` argument in Contract methods
### Feature Request

Argument `block_number` has been a part of Client interface for a good while. Contract methods should be updated to include this argument alongside the `block_hash`.

### PR DESCRIPTION
**Pull request by drknzz**
Closes #823 

It would be best to have both `block_number` and `block_hash` as *keyword only* parameters, however it would be breaking.

### CODE REVIEW

File: starknet_py/contract.py [hunk 1]
```diff
@@ -256,26 +256,32 @@ def _account(self) -> BaseAccount:
...
         """
-        return await self._client.call_contract(call=self, block_hash=block_hash)
+        return await self._client.call_contract(
+            call=self, block_hash=block_hash, block_number=block_number
+        )
 
     async def call(
         self,
         block_hash: Optional[str] = None,
+        block_number: Optional[Union[int, Tag]] = None,
```
Comment (cptartur) [inline starknet_py/contract.py:275]: What if we made at least `block_number` a keyword only one? 

Ideally there'd be a way to detect if `block_hash` was passed as a positional argument and show a warning, but that might not be possible.

File: starknet_py/tests/e2e/client/client_test.py [hunk 2]
```diff
@@ -167,7 +167,7 @@ async def test_call_contract(client, contract_address):
         calldata=[],
     )
 
-    result = await client.call_contract(call, block_hash="latest")
+    result = await client.call_contract(call, block_number="latest")
```
Comment (cptartur) [inline starknet_py/tests/e2e/client/client_test.py:170]: Why was this changed?

File: starknet_py/contract.py (hunk 1 above)
Comment (drknzz) [reply starknet_py/contract.py:275]: I'd rather keep both of them simple, non keyword only / both keyword.

File: starknet_py/tests/e2e/client/client_test.py (hunk 2 above)
Comment (drknzz) [reply starknet_py/tests/e2e/client/client_test.py:170]: To avoid confusion, as `latest` / `pending` is not a valid `block_hash`.

Comment (cptartur) [reply starknet_py/tests/e2e/client/client_test.py:170]: It's still valid in terms of `Client` interface, but I'm fine with the change.

**Review by cptartur — APPROVED** [review]
